### PR TITLE
MediaTrackList: Fix crash in indexer binding code

### PR DESCRIPTION
**Pull request by mfkl**
https://code.videolan.org/videolan/vlc-unity/-/issues/104

https://mailman.videolan.org/pipermail/vlc-devel/2021-April/143192.html

### CODE REVIEW

File: src/LibVLCSharp/Structures/MediaTrack.cs [hunk 1]
```diff
@@ -284,7 +284,8 @@ internal MediaTrackList(IntPtr mediaTrackListPtr) : base(() => mediaTrackListPtr
         {
             get
             {
-                if (position > Count)
+                var count = Count;
+                if (count == 0 || position > count)
```
Comment (jeremyVignelles) [inline src/LibVLCSharp/Structures/MediaTrack.cs:288]: Should likely be `if (position >= count)`. What if Count is 1 and `[1]` is called?

Comment (mfkl) [reply src/LibVLCSharp/Structures/MediaTrack.cs:288]: > What if Count is 1 and `[1]` is called?

Hmm then that'd be an invalid call I guess? Count is 1-based while the indexer (`libvlc_media_tracklist_at`) is 0-based. Making me doubt now :D

Comment (mfkl) [reply src/LibVLCSharp/Structures/MediaTrack.cs]: Added your suggestion just to be safe, though from my understanding that condition would never have ran considering the following list implementation
```
public bool MoveNext()
{
    position++;
    return position < (_mediaTrackList?.Count ?? 0);
}
```

File: src/LibVLCSharp/Structures/MediaTrack.cs [hunk 2]
```diff
@@ -284,7 +284,8 @@ internal MediaTrackList(IntPtr mediaTrackListPtr) : base(() => mediaTrackListPtr
         {
             get
             {
-                if (position > Count)
+                var count = Count;
+                if (count == 0 || position >= count)
```
Comment (jeremyVignelles) [inline src/LibVLCSharp/Structures/MediaTrack.cs:288]: Could be simplified to the one-liner :
```suggestion
                if (position >= Count)
```

if `Count == 0` there's no way for an `uint` to be below

Comment (mfkl) [reply src/LibVLCSharp/Structures/MediaTrack.cs]: Sure

File: src/LibVLCSharp/Structures/MediaTrack.cs (hunk 1 above)
Comment (jeremyVignelles) [reply src/LibVLCSharp/Structures/MediaTrack.cs]: `MoveNext` is called when iterating in a `foreach` loop or when doing linq, but the indexer is not used in that case.

The proper test case would be to call `mediaTrackList[1]` directly, without checking the bounds.

**Review by jeremyVignelles — APPROVED** [review]

File: src/LibVLCSharp/Structures/MediaTrack.cs (hunk 1 above)
Comment (mfkl) [reply src/LibVLCSharp/Structures/MediaTrack.cs]: > MoveNext is called when iterating in a foreach loop or when doing linq, but the indexer is not used in that case.

Yes, it is (but it doesn't matter) :-) The `IEnumerator` states that after `MoveNext` is called when iterating over a list such as the media track list, `Current` is then called, which does `return _mediaTrackList[position]`.

**Pull request merged**